### PR TITLE
Fix dashboard description markdown overflow

### PR DESCRIPTION
**Pull request by deniskaber**
Closes https://github.com/metabase/metabase/issues/31326

### Description

There is an issue with markdown in dashboard description. It overflows its allocated area due to technical implementation. 
We decided to fix this.

### Demo

https://github.com/metabase/metabase/assets/7983744/1a917e84-b853-466b-bb3d-98044569dc60

### Checklist

- [x] Tests have been added/updated to cover changes in this PR

### CODE REVIEW

File: e2e/test/scenarios/dashboard/dashboard.cy.spec.js [hunk 1]
```diff
@@ -683,6 +683,45 @@ describe("scenarios > dashboard", () => {
...
+        expect($markdown[0].clientHeight).to.be.gte(
+          $markdown[0].firstElementChild.clientHeight,
+        );
+      });
+
+      cy.findByTestId("editable-text")
+        .click()
+        .then($el => {
+          expect($el[0].scrollHeight).to.be.gte(
+            testMarkdownContent.split("{enter}").length * 16, // num of lines * font size
```
Comment (kamilmielnik) [inline e2e/test/scenarios/dashboard/dashboard.cy.spec.js:720]: 1. I think it should be `numberOfLines * lineHeight` (+ padding?)
2. We don't need to hardcode it, we can use `parseFloat(window.getComputedStyle($el[0]).lineHeight)`

File: e2e/test/scenarios/dashboard/dashboard.cy.spec.js [hunk 2]
```diff
@@ -683,6 +683,45 @@ describe("scenarios > dashboard", () => {
       cy.findByText("There was a problem displaying this chart.");
     });
   });
+
+  it("should now overflow markdown content in description (metabase#31326)", () => {
```
Comment (kamilmielnik) [inline e2e/test/scenarios/dashboard/dashboard.cy.spec.js:687]: typo?
```suggestion
  it("should not overflow markdown content in description (metabase#31326)", () => {
```

**Review by kamilmielnik — COMMENTED** [review]
LGTM. I'll let others approve for now though.

File: frontend/src/metabase/core/components/EditableText/EditableText.styled.tsx [hunk 3]
```diff
@@ -27,12 +27,16 @@ export const EditableTextRoot = styled.div<EditableTextRootProps>`
...
-    word-wrap: break-word;
-  }
+  ${({ isEditingMarkdown }) =>
+    isEditingMarkdown &&
+    css`
+      &:after {
+        content: attr(data-value);
+        visibility: hidden;
+        white-space: pre-wrap;
+        word-wrap: break-word;
```
Comment (WiNloSt) [inline frontend/src/metabase/core/components/EditableText/EditableText.styled.tsx:37]: I think we should add `word-wrap: break-word` to the container too. Otherwise, we will still experience the horizontal overflow.

This screenshot is from running this PR
image

File: e2e/test/scenarios/dashboard/dashboard.cy.spec.js (hunk 2 above)
Comment (WiNloSt) [inline e2e/test/scenarios/dashboard/dashboard.cy.spec.js:687]: I assume this was a typo
```suggestion
  it("should not overflow markdown content in description (metabase#31326)", () => {
```

File: e2e/test/scenarios/dashboard/dashboard.cy.spec.js [hunk 4]
```diff
@@ -683,6 +683,45 @@ describe("scenarios > dashboard", () => {
       cy.findByText("There was a problem displaying this chart.");
     });
   });
+
```
Comment (kamilmielnik) [inline e2e/test/scenarios/dashboard/dashboard.cy.spec.js:686]: I don't see a test for the [horizontal overflow](https://github.com/metabase/metabase/assets/6830683/abb42e48-fc8f-44cc-be0a-de68dfd4eeb0), but not sure if we should go into such detail.

File: e2e/test/scenarios/dashboard/dashboard.cy.spec.js [hunk 5]
```diff
@@ -683,6 +683,45 @@ describe("scenarios > dashboard", () => {
...
+  it("should now overflow markdown content in description (metabase#31326)", () => {
+    cy.intercept("GET", "/api/dashboard/1").as("getDashboard");
+    cy.intercept("PUT", "/api/dashboard/1").as("updateDashboard");
+    visitDashboard(1);
+    cy.wait("@getDashboard");
+
+    cy.get("main header").within(() => {
+      cy.icon("info").click();
+    });
```
Comment (WiNloSt) [inline e2e/test/scenarios/dashboard/dashboard.cy.spec.js:695]: nit: We could use [`icon`](https://github.com/metabase/metabase/blob/master/e2e/support/commands/ui/icon.js) right after a subject now

```suggestion
    cy.get("main header").icon("info").click();
```

File: e2e/test/scenarios/dashboard/dashboard.cy.spec.js
Comment (WiNloSt) [inline e2e/test/scenarios/dashboard/dashboard.cy.spec.js:1]: It's still missing test(s) for Horizontal overflow

File: e2e/test/scenarios/dashboard/dashboard.cy.spec.js [hunk 6]
```diff
@@ -683,6 +683,45 @@ describe("scenarios > dashboard", () => {
...
+
+    const testMarkdownContent =
+      "{selectall}# Heading 1{enter}{enter}**bold** [link](https://metabase.com){enter}{enter}![alt](https://upload.wikimedia.org/wikipedia/commons/a/a2/Cat_outside.jpg){enter}{enter}This is my description. ";
```
Comment (kamilmielnik) [inline e2e/test/scenarios/dashboard/dashboard.cy.spec.js:698]: If the link to this image dies, it will affect this test.

Can we make this test independent of 3rd parties?

Comment (WiNloSt) [reply e2e/test/scenarios/dashboard/dashboard.cy.spec.js:698]: We can choose some static images from `resources/frontend_client/app/assets/img/` e.g. `resources/frontend_client/app/assets/img/welcome-modal-2.png` which is 
image

File: e2e/test/scenarios/dashboard/dashboard.cy.spec.js [hunk 7]
```diff
@@ -683,6 +683,59 @@ describe("scenarios > dashboard", () => {
...
+
+    cy.wait("@updateDashboard");
+
+    rightSidebar().within(() => {
+      // check that markdown content is not bigger than its container
+      cy.findByTestId("editable-text").then($markdown => {
+        const el = $markdown[0];
+
+        // vertical
+        expect($markdown[0].clientHeight).to.be.gte(
```
Comment (ranquild) [inline e2e/test/scenarios/dashboard/dashboard.cy.spec.js:713]: could we rely on elements being visible / hidden? It's hard to parse clientHeight checks

**Review by ranquild — APPROVED** [review]

File: e2e/test/scenarios/dashboard/dashboard.cy.spec.js [hunk 8]
```diff
@@ -683,6 +683,59 @@ describe("scenarios > dashboard", () => {
...
+          $markdown[0].firstElementChild.clientHeight,
+        );
+
+        // horizontal
+        $markdown.find("*").each((index, childEl) => {
+          expect(el.clientWidth).to.be.gte(
+            childEl.getBoundingClientRect().width,
+          );
+        });
```
Comment (kamilmielnik) [inline e2e/test/scenarios/dashboard/dashboard.cy.spec.js:722]: A child can overflow its parent even if the parent is wider. Maybe this is isn't true now, but styling could change.
To make this assertion more future-proof we can do this:

```tsx
const childRect = childEl.getBoundingClientRect();
const parentRect = $markdown[0].getBoundingClientRect();

expect(parentRect.left).to.be.lte(childRect.left);
expect(parentRect.right).to.be.gte(childRect.right);
```

**Review by kamilmielnik — COMMENTED** [review]
LGTM. Left 1 comment about making an assertion more future-proof.

File: frontend/src/metabase/core/components/EditableText/EditableText.styled.tsx [hunk 9]
```diff
@@ -27,12 +28,16 @@ export const EditableTextRoot = styled.div<EditableTextRootProps>`
...
-    content: attr(data-value);
-    visibility: hidden;
-    white-space: pre-wrap;
-    word-wrap: break-word;
-  }
+  ${({ isEditingMarkdown }) =>
+    isEditingMarkdown &&
+    css`
+      &:after {
+        content: attr(data-value);
```
Comment (uladzimirdev) [inline frontend/src/metabase/core/components/EditableText/EditableText.styled.tsx:35]: couldn't we just set `content` dynamically based on the isEditingMarkdown property?

**Review by WiNloSt — APPROVED** [review]
Awesome the test seems to work great!
